File: campus-navigate-app/expo-app/src/tiled/model/pathSegmentModel.ts
import type { OverlayPathSegment, OverlayPoint } from '../type';
import { distanceBetweenPoints } from './geometryMath';

export function createPathSegmentsFromPoints(
  points: ReadonlyArray<OverlayPoint & { nodeId?: string }>,
) {
  const segments: OverlayPathSegment[] = [];

  for (let index = 0; index < points.length - 1; index += 1) {
    const from = points[index];
    const to = points[index + 1];
    if (!from || !to || distanceBetweenPoints(from, to) <= 0.001) {
      continue;
    }

    const fromId = from.nodeId ?? `point-${index}`;
    const toId = to.nodeId ?? `point-${index + 1}`;
    segments.push(createPathSegment(`${fromId}->${toId}`, from, to, fromId, toId));
  }

  return segments;
}

function createPathSegment(
  key: string,
  from: OverlayPoint,
  to: OverlayPoint,
  fromNodeId = key.split('->')[0] ?? 'from',
  toNodeId = key.split('->')[1] ?? 'to',
): OverlayPathSegment {
  const deltaX = to.screenX - from.screenX;
  const deltaY = to.screenY - from.screenY;

  return {
    fromNodeId,
    key,
    length: Math.hypot(deltaX, deltaY),
    rotationDegrees: (Math.atan2(deltaY, deltaX) * 180) / Math.PI,
    toNodeId,
    x: from.screenX,
    y: from.screenY,
  };
}

File: tarumt-nav-app/flutter_app/lib/domain/tiled/map/path_segment_model.dart
import 'package:indoor_navigation/domain/common/geometry_math.dart';
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

List<OverlayPathSegment> createPathSegmentsFromPoints(
  List<OverlayPoint> points,
) {
  final segments = <OverlayPathSegment>[];

  for (var index = 0; index < points.length - 1; index += 1) {
    final from = points[index];
    final to = points[index + 1];
    if (distanceBetweenPoints(from, to) <= 0.001) {
      continue;
    }

    final fromId = from is OverlayRouteNode ? from.nodeId : 'point-$index';
    final toId = to is OverlayRouteNode ? to.nodeId : 'point-${index + 1}';
    segments.add(_createPathSegment('$fromId->$toId', from, to, fromId, toId));
  }

  return segments;
}

OverlayPathSegment _createPathSegment(
  String key,
  OverlayPoint from,
  OverlayPoint to,
  String fromNodeId,
  String toNodeId,
) {
  return OverlayPathSegment(
    fromNodeId: fromNodeId,
    key: key,
    length: distanceBetweenPoints(from, to),
    rotationDegrees: headingBetweenPoints(from, to),
    toNodeId: toNodeId,
    x: from.screenX,
    y: from.screenY,
  );
}
